lib/playlist-generation/synthesize-playlist.ts

```diff
@@ -0,0 +1,69 @@
+import { AudioFeatures, Song, User } from '..';
+import { getNRandomElements, kLargest } from '../../utils';
+import { Lobby } from '../lobby';
+import { addSongs } from '../spotify/add-songs';
+import { THEME, theme2Conditions } from './themes';
+
+/**
+ *
+ * @param lobbyId the lobby to analyze
+ */
+export async function synthesizePlaylist(lobbyId: string, theme: THEME): Promise<boolean> {
+  const lobby = await Lobby.fromId(lobbyId);
+  if (!lobby) return false;
+
+  const songsMap = await lobby.participants.reduce(async (accP: Promise<Record<string, number>>, userId) => {
+    const acc = await accP;
+
+    const user = await User.fromId(userId);
+    if (!user) return acc;
+
+    user.topSongs.forEach((songId) => {
+      if (!acc[songId]) acc[songId] = 0;
+      acc[songId]++;
+    });
+
+    return acc;
+  }, Promise.resolve({}));
+
+  const songScores: Song2Score[] = [];
+
+  Object.entries(songsMap).forEach(async ([id, overlapScore]) => {
+    const song = await Song.fromId(id);
+    if (!song || !song.audioFeatures) return;
+    const score = computeScore(song.audioFeatures, theme) * (1 + overlapScore * .1);
+    if (score === 0) return;
+    songScores.push({id, score});
+  });
+
+  const topSongs = kLargest<Song2Score>(songScores, compareSongScores, 50);
+
+  const additionalSongs = topSongs.length === 50
+    ? []
+    : getNRandomElements(Object.keys(songsMap), 50 - topSongs.length);
```

Comment:
will this potentially grab duplicate songs? i think spotify lets u add the same song twice to a playlist

Reply:
hm yes.. maybe thats a problem for another day HAHA